pytest_leaks/refleak_35.py

```diff
@@ -0,0 +1,281 @@
+        alloc_before, rc_before = alloc_after, rc_after
+    #print(file=sys.stderr)  # <- pytest-leaks edit
+
+    # These checkers return False on success, True on failure
+    def check_rc_deltas(deltas):
+        # Checker for reference counters and memomry blocks.
+        #
+        # bpo-30776: Try to ignore false positives:
+        #
+        #   [3, 0, 0]
+        #   [0, 1, 0]
+        #   [8, -8, 1]
+        #
+        # Expected leaks:
+        #
+        #   [5, 5, 6]
+        #   [10, 1, 1]
+        return all(delta >= 1 for delta in deltas)
+
+    failed = False
+    leaks = OrderedDict()  # <- pytest-leaks edit
+    for deltas, item_name, checker in [
+        (rc_deltas, 'references', check_rc_deltas),
+        (alloc_deltas, 'memory blocks', check_rc_deltas)
+    ]:
+        # ignore warmup runs
+        deltas = deltas[nwarmup:]
+        if checker(deltas):
+            # <pytest-leaks edit>
+            leaks[item_name] = deltas
+            continue
+            # </pytest-leaks edit>
+            msg = '%s leaked %s %s, sum=%s' % (
+                test, deltas, item_name, sum(deltas))
+            print(msg, file=sys.stderr)
+            sys.stderr.flush()
+            with open(fname, "a") as refrep:
+                print(msg, file=refrep)
+                refrep.flush()
+            failed = True
+    return leaks  # <- pytest-leaks edit
+
+def dash_R_cleanup(fs, ps, pic, zdc, abcs):
+    import gc, copyreg
+    import collections.abc
+    from weakref import WeakSet
+
+    # Restore some original values.
+    warnings.filters[:] = fs
+    copyreg.dispatch_table.clear()
+    copyreg.dispatch_table.update(ps)
+    sys.path_importer_cache.clear()
+    sys.path_importer_cache.update(pic)
+    try:
+        import zipimport
+    except ImportError:
+        pass # Run unmodified on platforms without zipimport support
+    else:
+        zipimport._zip_directory_cache.clear()
+        zipimport._zip_directory_cache.update(zdc)
+
+    # clear type cache
+    sys._clear_type_cache()
+
+    # Clear ABC registries, restoring previously saved ABC registries.
+    abs_classes = [getattr(collections.abc, a) for a in collections.abc.__all__]
+    abs_classes = filter(isabstract, abs_classes)
+    if 'typing' in sys.modules:
+        t = sys.modules['typing']
+        # these classes require special treatment because they do not appear
+        # in direct subclasses on collections.abc classes
+        abs_classes = list(abs_classes) + [t.ChainMap, t.Counter, t.DefaultDict]
+    for abc in abs_classes:
+        for obj in abc.__subclasses__() + [abc]:
+            obj._abc_registry = abcs.get(obj, WeakSet()).copy()
+            obj._abc_cache.clear()
+            obj._abc_negative_cache.clear()
+
+    clear_caches()
+
+    # Collect cyclic trash and read memory statistics immediately after.
+    func1 = sys.getallocatedblocks
+    func2 = sys.gettotalrefcount
+    gc.collect()
```

Comment:
I am curious if we should do `for i in range(20): if gc.collect() == 0: break` (possibly with 1-2 extra runs), or leave deep cyclic cleanup to the test writer (I do not care much either way, I had to do such a thing in numpy in one spot; And I suppose suppose it might slow down things a bit, since it will probably run at least twice usually, instead of once.)

Reply:
The cpython>=3.7 refleak code indeed calls gc multiple (3) times.

Here, I didn't really change the cpython code at all, it's direct copypaste. I guess we can rationalize this when we know better, or combine the codebases for the different Python versions.

Reply:
Ah ok, I did not realize. Breaking early if 0 is reached may be a nice thing. I think Matti mentioned that at least in pypy, you need 2-3 extra calls (OTOH, I do not think pypy is of primary concern here).